platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/AdditionalDataBuilderFactory.java

```diff
@@ -16,35 +16,116 @@
 
 package org.gradle.api.problems.internal;
 
+import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Map;
 
 public class AdditionalDataBuilderFactory {
 
+    static String getSupportedTypes() {
+        StringBuilder result = new StringBuilder();
+        for (Class<?> key : additionalDataBuilderProviders.keySet()) {
+            if (result.length() > 0) {
+                result.append(", ");
+            }
+            result.append(key.getName());
+        }
+        return result.toString();
+    }
+
+    static class DataTypeAndProvider {
+        final Class<?> dataType;
+        final Function<Object, AdditionalDataBuilder<? extends AdditionalData>> builderProvider;
+
+        DataTypeAndProvider(Class<?> dataType, Function<Object, AdditionalDataBuilder<? extends AdditionalData>> builderProvider) {
+            this.dataType = dataType;
+            this.builderProvider = builderProvider;
+        }
+    }
+
+    static Map<Class<?>, DataTypeAndProvider> additionalDataBuilderProviders = ImmutableMap.<Class<?>, DataTypeAndProvider>builder()
+        .put(GeneralDataSpec.class, new DataTypeAndProvider(
+            GeneralData.class,
+            new Function<Object, AdditionalDataBuilder<? extends AdditionalData>>() {
+                @Override
+                public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable Object instance) {
+                    if (instance == null) {
+                        return DefaultGeneralData.builder();
+                    }
+                    return DefaultGeneralData.builder((GeneralData) instance);
```

Comment:
❓ Couldn't these nullability checks abstracted away? AFAIK `null` is castable, so it won't cause a problem here.